### PR TITLE
perf(gatsby,gatsby-plugin-gatsby-cloud): prioritize content/asset fetching early, drop script preloads so js doesn't compete with assets

### PR DESCRIPTION
**Pull request by marvinjude**
## Description

In this PR:

We prioritised fetching assets like fonts and images. This was accomplished by removing both from JSON & javascript preload  in HTML and `gatsby-plugin-gatsby-cloud`.

We Delayed Javascript execution by loading only our app bundle, and other essential bits then lazily loading the page template. This moves hydration to a later stage while allowing the browser to prioritise fonts and images

## Todo

- [x] Update tests

[ch48769]
[ch48768]
[ch48776]

### CODE REVIEW

**Comment by axe312ger** [comment]
Cool!

File: packages/gatsby/src/utils/client-assets-for-template.ts [hunk 1]
```diff
@@ -63,13 +63,21 @@ export async function getScriptsAndStylesForTemplate(
     }
   }
 
+  // console.log(componentChunkName)
+
```
Comment (pieh) [inline packages/gatsby/src/utils/client-assets-for-template.ts:67]:
```suggestion
```

File: packages/gatsby/src/utils/client-assets-for-template.ts [hunk 2]
```diff
@@ -63,13 +63,21 @@ export async function getScriptsAndStylesForTemplate(
...
+    let assets = webpackStats.assetsByChunkName[chunkName]
+
     if (!assets) {
       continue
     }
 
+    // Remove JS asset for templates
+    if (chunkName !== `app`) {
+      assets = assets.filter(asset => !asset.endsWith(`.js`))
+    }
```
Comment (pieh) [inline packages/gatsby/src/utils/client-assets-for-template.ts:77]: This removes "top level" assets, but below we also have handling for "webpack magic comments" and we should filter out `.js` out as well there I think (just for page template ones, not for the `app`)?

Comment (marvinjude) [reply packages/gatsby/src/utils/client-assets-for-template.ts]: Oh yes! Good catch

File: packages/gatsby/src/utils/client-assets-for-template.ts [hunk 3]
```diff
@@ -98,14 +97,24 @@ export async function getScriptsAndStylesForTemplate(
...
+
     const childAssets = webpackStats.childAssetsByChunkName[chunkName]
     if (!childAssets) {
       continue
     }
 
     for (const [rel, assets] of Object.entries(childAssets)) {
+      let childNonTemplateJSAssets
+      // Remove JS asset for templates
+      if (rel !== `app`) {
```
Comment (pieh) [inline packages/gatsby/src/utils/client-assets-for-template.ts:109]: rel would be `preload` or `prefetch`, we should check `chunkName` here

Comment (marvinjude) [reply packages/gatsby/src/utils/client-assets-for-template.ts]: 👍

File: e2e-tests/path-prefix/cypress/integration/asset-prefix.js [hunk 4]
```diff
@@ -11,10 +11,6 @@ describe(`assetPrefix`, () => {
   })
 
   describe(`runtime`, () => {
-    it(`prefixes preloads`, () => {
```
Comment (wardpeet) [inline e2e-tests/path-prefix/cypress/integration/asset-prefix.js:14]: should we check if no preloads exist?

Comment (marvinjude) [reply e2e-tests/path-prefix/cypress/integration/asset-prefix.js:14]: Sure! That'd be good to have

Comment (marvinjude) [reply e2e-tests/path-prefix/cypress/integration/asset-prefix.js:14]: @wardpeet We are only removing non "app" bundle JavaScript. Do you mean to test if we only have webpack-runtime app etc?

Comment (wardpeet) [reply e2e-tests/path-prefix/cypress/integration/asset-prefix.js:14]: we should still remove preloads from app.js

Comment (marvinjude) [reply e2e-tests/path-prefix/cypress/integration/asset-prefix.js:14]: 👍

**Pull request merged**